### PR TITLE
fix: Add a verity_cleanup function to test helpers.bash

### PR DESCRIPTION
**Pull request by raharper**
Add a cleanup function to unmount underlying atoms if one of the mount fails.

**What type of PR is this?**

bug

**Which issue does this PR fix**:

#541

**What does this PR do / Why do we need it**:

Add a cleanup function which tracks the successfully mounted atoms, and if one
fails, then it will unmount the successful atoms before returning the error.

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

make check locally.

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

no

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by smoser** [comment]
Do you have evidence or experience that this actually fixes the issue?

Also, you should 'Fixes #541' in your commit message.

**Comment by raharper** [comment]
> Do you have evidence or experience that this actually fixes the issue?

Not yet, I didn't have a setup to run the CI locally (going to set that up now); 

> 
> Also, you should 'Fixes #541' in your commit message.

Sure, will add.

**Comment by raharper** [comment]
```
ok 6 --no-squashfs-verity works in 2sec
ok 7 mount + umount works in 2sec
ok 8 mount + umount + mount a tree of images works in 3sec
ok 9 bad existing verity device is rejected in 12sec
```
\o/

**Review by smoser — CHANGES_REQUESTED** [review]
thank you for digging on this. i just ran 'losetup -a' on my laptop and found:
```
$ losetup -a | grep stackertest |wc -l
33
```
But only 9 of those 33 had 'verity' . 
there are several others.

```
$ losetup -a | sed -e '/stackertest/!d' -e 's,.*/stackertest-,,' -e 's,[.].*,,' | sort -u
test_bad_existing_verity_device_is_rejected
test_mount_-2b_umount_-2b_mount_a_tree_of_images_works
```

Do you know if this will also fix the 'mount a tree of images' ?

File: pkg/atomfs/molecule.go [hunk 1]
```diff
@@ -24,6 +24,19 @@ type Molecule struct {
...
+	// in the case that we have a verity or other mount error we need to
+	// tear down the other underlying atoms so we don't leave verity and loop
+	// devices around unused.
+	atomsMounted := []string{}
+	cleanupAtoms := func(err error) error {
+		for _, target := range atomsMounted {
+			if err := squashfs.Umount(target); err != nil {
+				return errors.Wrapf(err, "failed to unmount atom @ target %q while handling error: %s", target, err)
+			}
+		}
```
Comment (smoser) [inline pkg/atomfs/molecule.go:36]: huh. this "not covered" is interesting.

we *should* be getting coverage "credit" for the integration tests (make test) and I would have thought the bad-verity test would have pushed us down this failure path. Right?

Comment (raharper) [reply pkg/atomfs/molecule.go:36]: Just to confirm that we do get coverage, here's a debug version of the change:

```golang
diff --git a/pkg/atomfs/molecule.go b/pkg/atomfs/molecule.go
index 4797481..f9e88be 100644
--- a/pkg/atomfs/molecule.go
+++ b/pkg/atomfs/molecule.go
@@ -1,6 +1,7 @@
 package atomfs

 import (
+       "fmt"
        "os"
        "path"
        "path/filepath"
@@ -24,6 +25,18 @@ type Molecule struct {
 // mountUnderlyingAtoms mounts all the underlying atoms at
 // config.MountedAtomsPath().
 func (m Molecule) mountUnderlyingAtoms() error {
+       atomsMounted := []string{}
+       cleanup := func(err error) error {
+               fmt.Printf("\nATOMFS.mountUnder.Cleanup func called\n")
+               for _, target := range atomsMounted {
+                       fmt.Printf("\nmount of molecule failed, unmounting previously mounted atom @ %s\n", target)
+                       if uerr := squashfs.Umount(target); uerr != nil {
+                               return errors.Wrapf(uerr, "failed to unmount atom @ target %s", target)
+                       }
+               }
+               return err
+       }
+
        for _, a := range m.Atoms {
                target := m.config.MountedAtomsPath(a.Digest.Encoded())

@@ -58,8 +71,9 @@ func (m Molecule) mountUnderlyingAtoms() error {

                err = squashfs.Mount(m.config.AtomsPath(a.Digest.Encoded()), target, rootHash)
                if err != nil {
-                       return errors.Wrapf(err, "couldn't mount")
+                       return cleanup(err)
                }
+               atomsMounted = append(atomsMounted, target)
        }

        return nil

diff --git a/pkg/squashfs/verity.go b/pkg/squashfs/verity.go
index 388ca0b..024b004 100644
--- a/pkg/squashfs/verity.go
+++ b/pkg/squashfs/verity.go
@@ -438,11 +438,13 @@ func Umount(mountpoint string) error {
        if !found {
                return errors.Errorf("%s is not a mountpoint", mountpoint)
        }
+       fmt.Printf("atomfs.Umount: source:%s mountpoint:%s\n", theMount.Source, mountpoint)

        err = unix.Unmount(mountpoint, 0)
        if err != nil {
                return errors.Wrapf(err, "failed unmounting %v", mountpoint)
        }
+       fmt.Printf("atomfs.Umount: unmounted mp %s\n", mountpoint)

        if _, err := os.Stat(theMount.Source); err != nil {
                if os.IsNotExist(err) {
@@ -461,17 +463,22 @@ func Umount(mountpoint string) error {
                        return err
                }

+               fmt.Printf("atomfs.Umount: verity getting handle to loop device /dev/loop%v\n", deviceNo)
                loopDev := losetup.New(uint64(deviceNo), 0)
+
                // here, we don't have the loopback device any more (we detached it
                // above). the cryptsetup API allows us to pass NULL for the crypt
                // device, but go-cryptsetup doesn't have a way to initialize a NULL
                // crypt device short of making the struct by hand like this.
+               fmt.Printf("atomfs.Umount: deactivating verity device %s\n", theMount.Source)
                err = (&cryptsetup.Device{}).Deactivate(theMount.Source)
                if err != nil {
                        return errors.WithStack(err)
                }
+               fmt.Printf("atomfs.Umount: verity device %s deactivated\n", theMount.Source)

                // finally, kill the loop dev
+               fmt.Printf("atomfs.Umount: verity detaching loop device /dev/loop%v\n", deviceNo)
                err = loopDev.Detach()
                if err != nil {
                        return errors.Wrapf(err, "failed to detach loop dev for %v", theMount.Source)
```

I changed the atomfs.bats test to not eat the output of the mount command
so we can see the full output of the failure path.

```
running: bad_stacker internal-go atomfs mount test-squashfs mountpoint | grep "invalid root hash"
----
Debug mode:
stacker version v1.0.0-rc9-dirty
stacker subcommand: [/home/ubuntu/stacker/stacker --debug internal-go atomfs mount test-squashfs mountpoint]
about to mount {[{application/vnd.stacker.image.layer.squashfs+zstd+verity sha256:43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c 8192 [] map[io.stackeroci.stacker.squashfs_verity_root_hash:5233cb6dee7c60ceaee7b14b3e10751f50bdb6a349101b127907d46be04fff99] [] <nil> } {application/vnd.stacker.image.layer.squashfs+zstd+verity sha256:608a28ca4af147ebdb537893426993864d7ab8364395bb33e84150d3332d388d 2060288 [] map[io.stackeroci.stacker.squashfs_verity_root_hash:a5a7f9661eb8c1a3c65669f23b4f5dc070fefef509a7c097c910f2925996a9e6] [] <nil> }] {/home/ubuntu/stacker/stackertest-test_BadExistingVerityDeviceIsRejected.KRGL9f/oci /home/ubuntu/stacker/stackertest-test_BadExistingVerityDeviceIsRejected.KRGL9f/atomfs-metadata test-squashfs mountpoint true}}

ATOMFS.mountUnder.Cleanup func called

mount of molecule failed, unmounting previously mounted atom @ /home/ubuntu/stacker/stackertest-test_BadExistingVerityDeviceIsRejected.KRGL9f/atomfs-metadata/mounts/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c
atomfs.Umount: source:/dev/mapper/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c-verity mountpoint:/home/ubuntu/stacker/stackertest-test_BadExistingVerityDeviceIsRejected.KRGL9f/atomfs-metadata/mounts/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c
atomfs.Umount: unmounted mp /home/ubuntu/stacker/stackertest-test_BadExistingVerityDeviceIsRejected.KRGL9f/atomfs-metadata/mounts/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c
atomfs.Umount: verity getting handle to loop device /dev/loop2
atomfs.Umount: deactivating verity device /dev/mapper/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c-verity
atomfs.Umount: verity device /dev/mapper/43ca103fe7d3dcaa0defff7f7375502e6149625c121803764f2b41db2ded436c-verity deactivated
atomfs.Umount: verity detaching loop device /dev/loop2
error: invalid root hash for 608a28ca4af147ebdb537893426993864d7ab8364395bb33e84150d3332d388d-verity: sha256 (expected: a5a7f9661eb8c1a3c65669f23b4f5dc070fefef509a7c097c910f2925996a9e6)
stackerbuild.io/stacker/pkg/squashfs.ConfirmExistingVerityDeviceHash
	/stacker-tree/pkg/squashfs/verity.go:554
stackerbuild.io/stacker/pkg/squashfs.HostMount
	/stacker-tree/pkg/squashfs/verity.go:373
stackerbuild.io/stacker/pkg/squashfs.Mount
	/stacker-tree/pkg/squashfs/verity.go:282
stackerbuild.io/stacker/pkg/atomfs.Molecule.mountUnderlyingAtoms
	/stacker-tree/pkg/atomfs/molecule.go:72
stackerbuild.io/stacker/pkg/atomfs.Molecule.Mount
	/stacker-tree/pkg/atomfs/molecule.go:158
main.doAtomfsMount
	/stacker-tree/cmd/stacker/internal_go.go:200
github.com/urfave/cli/v2.(*Command).Run
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:273
github.com/urfave/cli/v2.(*Command).Run
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:266
github.com/urfave/cli/v2.(*Command).Run
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:266
github.com/urfave/cli/v2.(*Command).Run
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:266
github.com/urfave/cli/v2.(*App).RunContext
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:332
github.com/urfave/cli/v2.(*App).Run
	/stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:309
main.main
	/stacker-tree/cmd/stacker/main.go:328
runtime.main
	/usr/lib/go/src/runtime/proc.go:267
runtime.goexit
	/usr/lib/go/src/runtime/asm_amd64.s:1650
```

Comment (smoser) [reply pkg/atomfs/molecule.go:36]: @raharper  you did good. thank you for pushing there.

I mis-read the codecov report. It complained about some lines in your diff not being covered, and I just assumed it was *all* of the unmount-handling code.  But it was only a small portion of it. Sorry for the noise.  I'm really happy that this happened though, it shows that codecov is working and that your code got excercised (all but the error handling part). 

Maybe the screenshot will help explain what I missed.  I just assumed it was saying lines 31-35, but it was only saying 34-35. 

image

File: pkg/atomfs/molecule.go [hunk 2]
```diff
@@ -24,6 +24,19 @@ type Molecule struct {
...
 // config.MountedAtomsPath().
 func (m Molecule) mountUnderlyingAtoms() error {
+	// in the case that we have a verity or other mount error we need to
+	// tear down the other underlying atoms so we don't leave verity and loop
+	// devices around unused.
+	atomsMounted := []string{}
+	cleanupAtoms := func(err error) error {
+		for _, target := range atomsMounted {
+			if err := squashfs.Umount(target); err != nil {
+				return errors.Wrapf(err, "failed to unmount atom @ target %q while handling error: %s", target, err)
```
Comment (smoser) [inline pkg/atomfs/molecule.go:34]: i think you've stomped on 'err' passed in here. so your wrapped 'err' is the *new* err, not the one passed in.

You should do something like:

```go
   if umountErr := squashfs.Umount(target); umountErr != nil {
    return errors.Wrapf(err, "failed to unmount atom @ target %q while handling error: %s", target, umountErr)
   }
```

I think that is better, but then do you want to return on the error or just collect it and continue trying to unmount the other targets ?

**Review by smoser — COMMENTED** [review]
This does look great, but i have one more comment inline.  I'm hope that I'm not wrong again and go 0-for-2 today on reading this PR.

File: pkg/atomfs/molecule.go (hunk 2 above)
Comment (raharper) [reply pkg/atomfs/molecule.go:34]: Note, in the case of an unmount error, I'll wrap the umountErr, but include out the contents of the passed in err in the wrapped message.

```golang
if umountErr := squashfs.Umount(target); umountErr != nil {
    return errors.Wrapf(umountErr, "failed to unmount atom @ target %q while handling error: %s", target, err)
}
```

**Review by smoser — APPROVED** [review]
this looks good. thanks.

**Pull request merged**